### PR TITLE
[FLINK-20645][network] Fix corrupted unaligned checkpoints.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

With unusual barrier flow, checkpoints may miss certain buffers and become unrecoverable.

## Brief change log

- Fix ChannelStatePersister. 
- Improve UnalignedCheckpointITCase to detect data corruption faster.
- Adding more trace output for network and checkpoint buffers as shown in the following

```
9996 [forward (1/1)#0] TRACE org.apache.flink.runtime.io.network.logger.NetworkActionsLogger [] - LocalInputChannel#getNextBuffer Buffer{size=1996, hash=-1568575630}, seq 4, ChannelStatePersister(1 (COMPLETED)} @ InputChannelInfo{gateIdx=0, inputChannelIdx=0}
9996 [Channel state writer Source: source (1/1)#0] TRACE org.apache.flink.runtime.io.network.logger.NetworkActionsLogger [] - ChannelStateWriteRequest#writeOutput Buffer{size=20, hash=1432560496}, checkpoint 2 @ ResultSubpartitionInfo{partitionIdx=0, subPartitionIdx=0}
9996 [forward (1/1)#0] TRACE org.apache.flink.runtime.io.network.logger.NetworkActionsLogger [] - PipelinedSubpartition#pollBuffer Buffer{size=38, hash=-1801242596} @ ResultSubpartitionInfo{partitionIdx=0, subPartitionIdx=0}
9996 [forward (1/1)#0] TRACE org.apache.flink.runtime.io.network.logger.NetworkActionsLogger [] - LocalInputChannel#getNextBuffer Buffer{size=38, hash=-1801242596}, seq 5, ChannelStatePersister(2 (BARRIER_RECEIVED)} @ InputChannelInfo{gateIdx=0, inputChannelIdx=0}
9996 [Source: source (1/1)#0] INFO  org.apache.flink.test.checkpointing.UnalignedCheckpointTestBase [] - Snapshotted LongSplit{increment=1, nextNumber=199, numCompletedCheckpoints=1} @ 0 subtask (0 attempt)
```

## Verifying this change

Already covered by existing UnalignedControllerTest when assertion is added.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java [hunk 1]
```diff
@@ -95,16 +115,39 @@ protected void maybePersist(Buffer buffer) {
     protected Optional<Long> checkForBarrier(Buffer buffer) throws IOException {
         final AbstractEvent event = parseEvent(buffer);
         if (event instanceof CheckpointBarrier) {
-            if (((CheckpointBarrier) event).getId() >= lastSeenBarrier) {
+            final long barrierId = ((CheckpointBarrier) event).getId();
+            long expectedBarrierId =
+                    checkpointStatus == CheckpointStatus.COMPLETED
+                            ? lastSeenBarrier + 1
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java:121]: Good catch! :1st_place_medal: :)

**Review by rkhachatryan — APPROVED** [review]
Thanks a lot for finding and fixing this bug!
The change LGTM in general (approving).
I left a couple of questions and nits.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/logger/NetworkActionsLogger.java [hunk 2]
```diff
@@ -18,51 +18,89 @@
...
+            Class<?> clazz,
+            String action,
+            Buffer buffer,
+            InputChannelInfo channelInfo,
+            ChannelStatePersister channelStatePersister,
+            int sequenceNumber) {
+        if (ENABLED) {
+            LOG.trace(
+                    "{}#{} {}, seq {}, {} @ {}",
+                    clazz.getSimpleName(),
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/logger/NetworkActionsLogger.java:46]: Good catch, I didn't know that `getSimpleName` is implemented without a cache. I have combined your idea of a string to also merge the action in it. So action now contains "{simpleName}#{method}" (or anything else that identifies what Flink does with the buffer).

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/logger/NetworkActionsLogger.java [hunk 3]
```diff
@@ -18,51 +18,89 @@
...
+                    action,
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/logger/NetworkActionsLogger.java:47]: I merged this parameter now with the class constant.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java [hunk 4]
```diff
@@ -62,9 +68,18 @@
     }
 
     protected void startPersisting(long barrierId, List<Buffer> knownBuffers) {
+        LOG.debug(
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java:71]: Yes, good idea. Without `channelInfo`, it's a bit useless. (Can be probably inferred in TRACE, but the idea was that these state transitions also appear in DEBUG)
Also I like the idea of extracting a log method.

**Pull request merged**